payload/orbit_determination/nonlinear_least_squares_od.py

```diff
@@ -0,0 +1,111 @@
+        :return: A numpy array of shape (6,) containing the next state (position and velocity).
+        """
+        return RK4(x, None, partial(self.state_derivative, epoch=epoch), self.dt)
+
+    def fit_orbit(self, starting_epoch: Epoch, times: np.ndarray, landmarks: np.ndarray, pixel_coordinates: np.ndarray,
+                  cubesat_attitudes: np.ndarray) -> np.ndarray:
+        """
+        Solve the orbit determination problem using non-linear least squares.
+
+        :param starting_epoch: The starting epoch as an instance of brahe's Epoch class.
+        :param times: A numpy array of shape (n,) containing the indices of time steps at which landmarks were observed.
+                      Must be sorted in ascending order.
+        :param landmarks: A numpy array of shape (n, 3) containing the ECI coordinates of the landmarks.
+        :param pixel_coordinates: A numpy array of shape (n, 2) containing the pixel coordinates of the landmarks.
+        :param cubesat_attitudes: A numpy array of shape (n, 4) containing the quaternions representing the attitude of the satellite.
+        :return: A numpy array of shape (max(times) + 1, 6) containing the ECI position and velocity of the satellite at each time step.
+        """
+        assert len(times.shape) == 1, "times must be a 1D array"
+        assert all(times[i] < times[i + 1] for i in range(len(times) - 1)), "times must be sorted in ascending order"
+        assert len(landmarks.shape) == 2, "landmarks must be a 2D array"
+        assert landmarks.shape[1] == 3, "landmarks must have 3 columns"
+        assert len(pixel_coordinates.shape) == 2, "pixel_coordinates must be a 2D array"
+        assert pixel_coordinates.shape[1] == 2, "pixel_coordinates must have 2 columns"
+        assert len(times) == len(landmarks) == len(pixel_coordinates), \
+            "times, landmarks, and pixel_coordinates must have the same length"
+
+        bearing_unit_vectors = np.zeros((len(times), 3))
+        for i in range(len(times)):
+            bearing_vector = self.camera.convert_pixel_coordinates_to_camera_ray_directions(pixel_coordinates[i, :])
+            bearing_unit_vectors[i] = bearing_vector / np.linalg.norm(bearing_vector)
+
+        eci_to_camera_rotations = np.zeros((len(times), 3, 3))
+        for i in range(len(times)):
+            eci_to_camera_rotations[i, ...] = self.camera.R_sat_cam @ Rotation.from_quat(q_inv(cubesat_attitudes[i, :]), scalar_first=True).as_matrix()
+
+        N = times[-1] + 1  # number of time steps
+
+        def residuals(X):
+            states = X.reshape(-1, 6)
+            positions = states[:, :3]
+
+            res = np.zeros(3 * (N - 1) + 3 * len(times))
+
+            # dynamics residuals
+            epoch = starting_epoch
+            for i in range(N - 1):
+                res[3 * i:3 * (i + 1)] = positions[i + 1] - self.f(states[i, :], epoch)[:3]
+                epoch = increment_epoch(epoch, self.dt)
```

Comment:
what's `epoch` being used for?

Reply:
I agree I don't think it should be necessary, but for some reason it's used in the [`gravity model`](https://github.com/cmu-argus-2/GNC-Simulation/blob/7ffc7a37cff94626ce1f5e88b2401cdd44cf2c85/world/physics/models/gravity.py#L32)

Reply:
But you're not using this instance of `epoch` in anything after incrementing it

Reply:
Oh wait nevermind, I see that it's being used in the dynamics